libp2p/src/builder/phase/tcp.rs

```diff
@@ -0,0 +1,208 @@
+                <SecUpgrade::Upgrade as InboundUpgrade<Negotiated<libp2p_tcp::$path::TcpStream>>>::Future: Send,
+                <SecUpgrade::Upgrade as OutboundUpgrade<Negotiated<libp2p_tcp::$path::TcpStream>>>::Future: Send,
+                <<<SecUpgrade as IntoSecurityUpgrade<libp2p_tcp::$path::TcpStream>>::Upgrade as UpgradeInfo>::InfoIter as IntoIterator>::IntoIter: Send,
+                <<SecUpgrade as IntoSecurityUpgrade<libp2p_tcp::$path::TcpStream>>::Upgrade as UpgradeInfo>::Info: Send,
+
+                MuxStream: StreamMuxer + Send + 'static,
+                MuxStream::Substream: Send + 'static,
+                MuxStream::Error: Send + Sync + 'static,
+                MuxUpgrade: IntoMultiplexerUpgrade<SecStream>,
+                MuxUpgrade::Upgrade: InboundUpgrade<Negotiated<SecStream>, Output = MuxStream, Error = MuxError> + OutboundUpgrade<Negotiated<SecStream>, Output = MuxStream, Error = MuxError> + Clone + Send + 'static,
+                <MuxUpgrade::Upgrade as InboundUpgrade<Negotiated<SecStream>>>::Future: Send,
+                <MuxUpgrade::Upgrade as OutboundUpgrade<Negotiated<SecStream>>>::Future: Send,
+                MuxError: std::error::Error + Send + Sync + 'static,
+                <<<MuxUpgrade as IntoMultiplexerUpgrade<SecStream>>::Upgrade as UpgradeInfo>::InfoIter as IntoIterator>::IntoIter: Send,
+                <<MuxUpgrade as IntoMultiplexerUpgrade<SecStream>>::Upgrade as UpgradeInfo>::Info: Send,
+            {
+                Ok(SwarmBuilder {
+                    phase: QuicPhase {
+                        transport: libp2p_tcp::$path::Transport::new(tcp_config)
+                            .upgrade(libp2p_core::upgrade::Version::V1Lazy)
+                            .authenticate(
+                                security_upgrade.into_security_upgrade(&self.keypair)?,
+                            )
+                            .multiplex(multiplexer_upgrade.into_multiplexer_upgrade())
+                            .map(|(p, c), _| (p, StreamMuxerBox::new(c))),
+                    },
+                    keypair: self.keypair,
+                    phantom: PhantomData,
+                })
+            }
+        }
+    };
+}
+
+impl_tcp_builder!("async-std", super::provider::AsyncStd, async_io);
+impl_tcp_builder!("tokio", super::provider::Tokio, tokio);
+
+impl<Provider> SwarmBuilder<Provider, TcpPhase> {
+    pub(crate) fn without_tcp(
+        self,
+    ) -> SwarmBuilder<Provider, QuicPhase<impl AuthenticatedMultiplexedTransport>> {
+        SwarmBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+            phase: QuicPhase {
+                // Note that this allows a user to build a faulty (i.e. dummy) transport only.
+                // Ideally the final build step would not exist unless one adds another transport
+                // after `without_tcp`. Though currently this would introduce significant
+                // complexity to each following step.
```

Comment:
Given that this function is crate-private, does this comment still hold value?

Reply:
Good point. It is used in the various shortcut methods. Though those themselves enforce the user to provide a valid transport.